modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/HistoricalRebalanceCheckpointTest.java

```diff
@@ -0,0 +1,386 @@
+     */
+    private void doTestDelayedToBackupsRequests(int nodes, boolean putAfterGaps) throws Exception {
+        int updateCnt = 2_000;
+
+        int backupNodes = prepareCluster(nodes, updateCnt);
+
+        Ignite prim = primaryNode(0L, DEFAULT_CACHE_NAME);
+
+        List<Ignite> backups = backupNodes(0L, DEFAULT_CACHE_NAME);
+
+        AtomicBoolean prepareBlock = new AtomicBoolean();
+        AtomicBoolean finishBlock = new AtomicBoolean();
+
+        AtomicReference<CountDownLatch> blockLatch = new AtomicReference<>();
+
+        TestRecordingCommunicationSpi.spi(prim).blockMessages(new IgniteBiPredicate<ClusterNode, Message>() {
+            @Override public boolean apply(ClusterNode node, Message msg) {
+                if ((msg instanceof GridDhtTxPrepareRequest && prepareBlock.get()) ||
+                    (msg instanceof GridDhtTxFinishRequest && finishBlock.get())) {
+                    CountDownLatch latch = blockLatch.get();
+
+                    assertTrue(latch.getCount() > 0);
+
+                    latch.countDown();
+
+                    return true;
+                }
+                else
+                    return false;
+            }
+        });
+
+        IgniteCache<Integer, Integer> primCache = prim.cache(DEFAULT_CACHE_NAME);
+
+        Consumer<Integer> cachePutAsync = (key) -> GridTestUtils.runAsync(() -> primCache.put(key, key));
+
+        try {
+            // Blocked at primary and backups.
+            prepareBlock.set(true);
+
+            blockLatch.set(new CountDownLatch(backupNodes * 20));
+
+            for (int i = 0; i < 20; i++)
+                cachePutAsync.accept(++updateCnt);
+
+            blockLatch.get().await();
+        }
+        finally {
+            prepareBlock.set(false);
+        }
+
+        if (backupNodes > 1) {
+            try {
+                // Blocked at backups only.
+                finishBlock.set(true);
+
+                blockLatch.set(new CountDownLatch(backupNodes * 30));
+
+                for (int i = 0; i < 30; i++)
+                    cachePutAsync.accept(++updateCnt);
+
+                blockLatch.get().await();
+            }
+            finally {
+                finishBlock.set(false);
+            }
+        }
+
+        if (putAfterGaps) {
+            for (int i = 0; i < 50; i++)
+                prim.cache(DEFAULT_CACHE_NAME).put(++updateCnt, updateCnt);
+        }
+
+        // Storing counters on primary.
+        forceCheckpoint();
+
+        // Emulating power off, OOM or disk overflow. Keeping data as is, with missed counters updates.
+        backups.forEach(node -> ((BlockableFileIOFactory)node.configuration().getDataStorageConfiguration()
+            .getFileIOFactory()).blocked = true);
+
+        List<String> backNames = backups.stream().map(Ignite::name).collect(Collectors.toList());
+
+        CountDownLatch rebalanceFinished = new CountDownLatch(1);
```

Comment:
You should make sure that it was a historical rebalance.

Use smth like 
```
LogListener lsnrRebalanceType = matches("fullPartitions=[], " +
            "histPartitions=[0]").times(backupNodes).build();
```

where 0 is a patrition number

Reply:
Fixed